test/src/DFS.idr

```diff
@@ -0,0 +1,88 @@
+module DFS
+
+import Data.Graph.Indexed.Query.DFS
+import Test.Data.Graph.Indexed.Generators
+import Text.Smiles.Simple
+import Data.SnocList
+
+%default total
+
+
+--------------------------------------------------------------------------------
+-- Example Graphs
+--------------------------------------------------------------------------------
+
+benzylbenzoate : Maybe (Graph Bond Elem)
+benzylbenzoate = readSmiles "C1=CC=C(C=C1)COC(=O)C2=CC=CC=C2"
+
+-- create a list with all nodes reachable from the start node and ignoring
+-- taboo nodes and their branches
+limitedDfsWithTest :
+     (taboo : List Nat)
+  -> Maybe (Graph e n)
+  -> (start : Nat)
+  -> List Nat
+  -> Bool
+limitedDfsWithTest _  Nothing        _ _  = False
+limitedDfsWithTest ts (Just $ G o g) s ns =
+  let Just b  := natToFin s o | Nothing => False
+      tsFin   := mapMaybe tryNatToFin ts
+      dfsWith := limitedDfsWith
+        {a=List (Fin o)}
+        g
+        tsFin
+        (\xs,x => Left (x :: xs))
+        Prelude.Nil
+        b
+   in case dfsWith of
+        Right _ => False
+        Left ls => reverse (map finToNat ls) == ns
```

Comment:
When you build up a list just to invert it at the end, you should be using a `SnocList`. The above can be simplified to:

```idris
  let Just b  := natToFin s o | Nothing => False
      tsFin   := mapMaybe tryNatToFin ts
   in map cast (limitedDfs g tsFin b <>> []) == ns
```